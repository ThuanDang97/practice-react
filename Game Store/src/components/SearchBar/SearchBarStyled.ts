import { styled } from '@mui/material'

export const SearchBarStyled = styled('input')`
  font-size: 14px;
  border: none;
  background-color: ${({ theme }) => theme.palette.grey[800]};
  color: ${({ theme }) => theme.palette.common.white};
  border-radius: 8px;
  padding: 12px 15px;
  display: flex;
  justify-content: flex-start;
  min-width: 350px;

  &:focus {
    outline: none;
  }
`
