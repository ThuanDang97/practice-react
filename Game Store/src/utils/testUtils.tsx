import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import { BrowserRouter as Router } from 'react-router-dom'

// Themes
import theme from '@self-types/Themes'

// Contexts
import { AuthProvider } from '@contexts/AuthProvider'
import { CartProvider } from '@contexts/CartProvider'

export const customRender = (component: ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <AuthProvider>
        <CartProvider>
          <Router>{component}</Router>
        </CartProvider>
      </AuthProvider>
    </ThemeProvider>,
  )
}

export * from '@testing-library/react'
export { customRender as render }
