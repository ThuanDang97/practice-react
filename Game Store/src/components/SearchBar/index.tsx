import { ChangeEvent } from 'react'

// Components
import { Search } from '@mui/icons-material'
import { SearchBarStyled } from '@components/SearchBar/SearchBarStyled'
import { Box, IconButton, useTheme } from '@mui/material'

export interface ISearchBarProps {
  onchange?: (e: ChangeEvent<HTMLInputElement>) => void
  onKeyDown: (e: React.KeyboardEvent<HTMLInputElement>) => void
  onClick: (e: React.MouseEvent<HTMLButtonElement>) => void
  value?: string
  placeholder?: string
}

const SearchBar = ({
  onchange,
  onKeyDown,
  onClick,
  value,
  placeholder,
}: ISearchBarProps) => {
  const theme = useTheme()

  return (
    <Box display="flex" alignItems="center">
      <SearchBarStyled
        type="text"
        onChange={onchange}
        onKeyDown={onKeyDown}
        placeholder={placeholder}
        value={value}
      />
      <IconButton type="submit" onClick={onClick}>
        <Search sx={{ color: theme.palette.background.paper }} />
      </IconButton>
    </Box>
  )
}

export default SearchBar
