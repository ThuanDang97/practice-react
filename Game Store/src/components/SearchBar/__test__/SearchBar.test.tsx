import { fireEvent, screen } from '@testing-library/react'

// Utils
import { render } from '@utils/testUtils'

// Components
import SearchBar from '@components/SearchBar'

describe('SearchBar component', () => {
  const handleChange = jest.fn()
  const handleKeyDown = jest.fn()
  const handleClick = jest.fn()

  it('should render Search', () => {
    const { container } = render(
      <SearchBar
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        onchange={handleChange}
        placeholder="Search games..."
      />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should call onChange Search ', () => {
    render(
      <SearchBar
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        onchange={handleChange}
        placeholder="Search games..."
      />,
    )

    const search = screen.getByRole('textbox')

    fireEvent.change(search, {
      target: { value: 'Search Value' },
    })

    expect(handleChange).toBeCalled()
    expect(search).toHaveValue('Search Value')
  })

  it('should call onClick Icon Search ', () => {
    render(
      <SearchBar
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        onchange={handleChange}
        placeholder="Search games..."
      />,
    )

    const iconSearch = screen.getByRole('button')

    fireEvent.click(iconSearch)
    expect(handleClick).toBeCalled()
  })

  it('should call onKeyDown after Enter ', () => {
    render(
      <SearchBar
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        onchange={handleChange}
        placeholder="Search games..."
      />,
    )

    const search = screen.getByRole('textbox')
    fireEvent.keyDown(search)

    expect(handleKeyDown).toBeCalled()
  })
})
