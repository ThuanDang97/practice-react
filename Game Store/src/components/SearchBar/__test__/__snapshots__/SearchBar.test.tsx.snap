// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component should render Search 1`] = `
<div>
  <div
    class="MuiBox-root css-70qvj9"
  >
    <input
      class="css-1innor3"
      placeholder="Search games..."
      type="text"
      value=""
    />
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="submit"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17slstj-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
